Gordon360/Services/ProfileService.cs

```diff
@@ -267,6 +278,114 @@ public async Task UpdateCustomProfileAsync(string username, string type, CUSTOM_
+        bool viewerIsPolice = viewerGroups.Contains(AuthGroup.Police);
+        bool viewerIsFacStaff = viewerGroups.Contains(AuthGroup.FacStaff);
+        bool viewerIsStudent = viewerGroups.Contains(AuthGroup.Student);
+        bool viewerIsAlumni = viewerGroups.Contains(AuthGroup.Alumni);
+        bool profileIsFacStaff = restricted_profile.PersonType.Contains(FACSTAFF_PROFILE);
+        bool profileIsStudent = restricted_profile.PersonType.Contains(STUDENT_PROFILE);
+        bool profileIsAlumni = restricted_profile.PersonType.Contains(ALUMNI_PROFILE);
+
+        foreach (UserPrivacy_Settings row in privacy)
+        {
+            if ((viewerIsSiteAdmin || viewerIsPolice) && row.Visibility != "Public")
+            {
+                MarkAsPrivate(restricted_profile, row.Field);
+            }
+            else if (viewerIsFacStaff)
+            {
+                if (profileIsFacStaff && row.Visibility == "Private")
+                {
+                    MakePrivate(restricted_profile, row.Field);
+                }
+                else if ((profileIsStudent || profileIsAlumni) && row.Visibility != "Public")
+                {
+                    MarkAsPrivate(restricted_profile, row.Field);
+                }
+            }
+            else if (viewerIsStudent && row.Visibility != "Public")
+            {
+                MakePrivate(restricted_profile, row.Field);
+            }
+            else if (viewerIsAlumni && row.Visibility != "Public")
+            {
+                MakePrivate(restricted_profile, row.Field);
+            }
+        }
+
+        return restricted_profile;
+    }
+
+    /// <summary>
+    /// get privacy setting for particular user
+    /// </summary>
+    /// <param name="username">AD username</param>
+    /// <returns>all privacy setting of the particular user</returns>
+    public IEnumerable<UserPrivacyViewModel> GetPrivacySettingAsync(string username)
+    {
+        var account = accountService.GetAccountByUsername(username);
+
+        // select all privacy settings
+        var privacy = context.UserPrivacy_Settings.Where(up_s => up_s.gordon_id == account.GordonID).Select(up_s => (UserPrivacyViewModel)up_s);
+
+        return privacy;
+    }
+
+    /// <summary>
+    /// privacy setting of some piece of personal data for user.
+    /// </summary>
+    /// <param name="username">AD Username</param>
+    /// <param name="userPrivacy">User Privacy Update View Model</param>
+    public async Task UpdateUserPrivacyAsync(string username, UserPrivacyUpdateViewModel userPrivacy)
+    {
+        var account = accountService.GetAccountByUsername(username);
+        foreach (string field in userPrivacy.Field)
+        {
+            var user = context.UserPrivacy_Settings.FirstOrDefault(up_s => up_s.gordon_id == account.GordonID && up_s.Field == field);
+            if (user is null)
+            {
+                var privacy = new UserPrivacy_Settings
```

Comment:
I may be a little ignorant here, but why this isn't a constructor call?

Reply:
Maybe it's because UserPrivacy_Settings is auto-generated by EF Power Tools, and I don't know if we could have it generate a constructor, or could create one in another file (not autogenerated).  Maybe what's here is the best of the options.

Reply:
This is probably the best way to handle this.  There is no constructor for this type, and this is the only place we find `new UserPrivacy_Settings` in the codebase.